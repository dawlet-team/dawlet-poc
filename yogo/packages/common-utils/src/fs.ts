import *  as fs from 'fs'
import { join } from 'path'

export const filterUnrelatedFolders = (folders: string[]) => folders.filter(folder => !UNRELATED_FOLDERS.includes(folder))

export const fetchAvailableDawlets = () => {
  return filterUnrelatedFolders(readPackagesFolder())
}

export const readPackagesFolder = () => {
  return fs.readdirSync(join(process.cwd(), '../'))
}

export const getPackageJson = (folder: string) => {
  const path = join(process.cwd(), '../', folder, 'package.json')
  const file = fs.readFileSync(path, { encoding: 'utf-8' })
  const json = JSON.parse(file)
  return json
}

const UNRELATED_FOLDERS = ['common-utils', 'core', 'launcher', 'react-electron-webpack-config']