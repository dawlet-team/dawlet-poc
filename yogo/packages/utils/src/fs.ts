import *  as fs from 'fs'
import { join } from 'path'

export const pathToDawletsDir = join(process.cwd(), '../../dawlets')

export const pkgNameWithoutPrefix = (pkgName: string) => {
  if (pkgName.includes('@dawlet')) {
    return pkgName.replace("@dawlet/", "")
  }
  throw new Error(`${pkgName} is not a Dawlet package`)
}

export const fetchAvailableDawlets = () => {
  return readDawletsDir().map(retrieveDawletConfig)
}

export const readDawletsDir = () => {
  return fs.readdirSync(pathToDawletsDir)
}

export const retrieveDawletConfig = (folderPath: string): Dawlet.Config => {
  const packageJson = getPackageJson(folderPath)
  return {
    name: packageJson.name || folderPath,
    description: packageJson.description || "Dawlets are Awesome!",
    localizable: packageJson["dawlet"].localizable || false
  }
}

export const getPackageJson = (folder: string) => {
  const path = join(pathToDawletsDir, folder, 'package.json')
  const file = fs.readFileSync(path, { encoding: 'utf-8' })
  const json = JSON.parse(file)
  return json
}