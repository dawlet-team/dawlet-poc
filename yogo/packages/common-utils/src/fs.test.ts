import { fetchAvailableDawlets, readPackagesFolder, filterUnrelatedFolders, getPackageJson, retrieveDawletConfig } from './fs'
describe('fetchAvailableDawlets', () => {
  it('reads packages folder', () => {
    const dawlets = readPackagesFolder()
    expect(dawlets).toMatchSnapshot()
  })
  it('filters out unrelated folders', () => {
    const folders = ['algolet', 'midilet', 'common-utils', 'core', 'launcher', 'react-electron-webpack-config']
    const filtered = filterUnrelatedFolders(folders)
    expect(filtered).toEqual(['algolet', 'midilet'])
  })
  it('fetches available dawlets', () => {
    expect(fetchAvailableDawlets()).toMatchSnapshot()
  })
})

describe('retrieveDawletConfig', () => {
  it('gets dawlet config', () => {
    const algoletConfig = retrieveDawletConfig("algolet")
    expect(algoletConfig).toMatchSnapshot()
  })
})

describe('getPackageJson', () => {
  it('gets package json', () => {
    const json = getPackageJson("algolet")
    expect(json).toBeInstanceOf(Object)
  })
})