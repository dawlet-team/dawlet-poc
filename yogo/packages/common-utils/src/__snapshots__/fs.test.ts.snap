// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAvailableDawlets fetches available dawlets 1`] = `
Array [
  Object {
    "description": "Algolet lets you write your own custom logic to generate music.",
    "localizable": true,
    "name": "algolet",
  },
  Object {
    "description": "Sheetlet displays the music represented as a sheetmusic.",
    "localizable": true,
    "name": "sheetlet",
  },
]
`;

exports[`fetchAvailableDawlets reads dawlets folder 1`] = `
Array [
  "algolet",
  "sheetlet",
]
`;

exports[`retrieveDawletConfig gets dawlet config 1`] = `
Object {
  "description": "Algolet lets you write your own custom logic to generate music.",
  "localizable": true,
  "name": "algolet",
}
`;
